Chapter 1 Solution 3 - URLify is incorrect

The solution fails for the following test case `lqxwL9 2T5wSg1YKd9Z3iOFL___` (`_` is used to represent space at the end).

The response needs to be `lqxwL9%202T5wSg1YKd9Z3iOFL` but is instead 
```
AssertionError: Lists differ: ['l', 'l', 'q', 'x', 'w', 'L', '9', '%', '2', '0[82 chars] 'L'] != ['l', 'q', 'x', 'w', 'L', '9', '%', '2', '0', '2[77 chars] 'L']
```

Here's a simpler solution using list comprehension:
Assumes that excess spaces are always at the end
```
def urlify(url, length):
    return list(''.join(['%20' if char == ' ' else char for char in string[:length]]))
```
Not sure if such a solution would be acceptable in an interview, though I can't see why not. This is also faster than the original solution.

--- a/chapter_01/p03_urlify.py
+++ b/chapter_01/p03_urlify.py
@@ -6,7 +6,6 @@ def urlify_algo(string, length):
     """replace spaces with %20 and removes trailing spaces"""
     # convert to list because Python strings are immutable
     char_list = list(string)
-    string = ""
     new_index = len(char_list)
 
     for i in reversed(range(length)):
@@ -19,12 +18,12 @@ def urlify_algo(string, length):
             char_list[new_index - 1] = char_list[i]
             new_index -= 1
     # convert back to string
-    return string.join(char_list)
+    return "".join(char_list[new_index:])
 
 
 def urlify_pythonic(text, length):
     """solution using standard library"""
-    return text.rstrip().replace(" ", "%20")
+    return text.strip().replace(" ", "%20")
 
 
 class Test(unittest.TestCase):
@@ -32,7 +31,7 @@ class Test(unittest.TestCase):
 
     test_cases = [
         ("much ado about nothing      ", "much%20ado%20about%20nothing"),
-        ("Mr John Smith    ", "Mr%20John%20Smith"),
+        ("Mr John Smith       ", "Mr%20John%20Smith"),
     ]
     testable_functions = [urlify_algo, urlify_pythonic]
 

